<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="MendelGWAS.jl : This package performs standard Genome-Wide Association Studies.">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>MendelGWAS.jl</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/OpenMendel/MendelGWAS.jl">View on GitHub</a>

          <h1 id="project_title">MendelGWAS.jl</h1>
          <h2 id="project_tagline">This package performs standard Genome-Wide Association Studies.</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/OpenMendel/MendelGWAS.jl/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/OpenMendel/MendelGWAS.jl/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h3>
<a id="overview" class="anchor" href="#overview" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Overview</h3>

<p>Mendel GWAS is a component of the umbrella <a href="https://openmendel.github.io">OpenMendel</a> project. This analysis option performs a standard Genome-Wide Association Study (GWAS) to assess common genetic variants in unrelated individuals.</p>

<h3>
<a id="appropriate-problems-and-data-sets" class="anchor" href="#appropriate-problems-and-data-sets" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Appropriate Problems and Data Sets</h3>

<p>Mendel GWAS analysis input data is unrelated individuals genotyped at a large number of autosomal or X-linked SNPs. Mendel GWAS uses the compressed SNP data files. </p>

<h3>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Installation</h3>

<p><em>Note: Three OpenMendel packages - <a href="https://github.com/OpenMendel/SnpArrays.jl">SnpArrays</a>, <a href="https://github.com/OpenMendel/Search.jl">Search</a>, and <a href="https://github.com/OpenMendel/MendelBase.jl">MendelBase</a> must be installed before any Mendel analysis packages will run.</em></p>

<p>Within Julia, use the package manager to install Mendel:</p>

<pre><code>Pkg.clone("git@github.com:OpenMendel/MendelGWAS.jl.git")
</code></pre>

<p>This package supports Julia v0.4.</p>

<h3>
<a id="input-files" class="anchor" href="#input-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Input Files</h3>

<p>The MendelGWAS analysis package uses the following input files. Example input files can be found in the <a href="https://github.com/OpenMendel/MendelGWAS.jl/tree/master/docs">docs</a> subfolder of the MendelGWAS project.</p>

<ul>
<li>
<a href="#control-file">Control File</a>: Specifies the names of your data input and output files and any optional parameters (<em>keywords</em>) for the analysis. (For a list of common keywords, see <a href="https://openmendel.github.io/MendelBase.jl/#keywords-table">Keywords Table</a>).</li>
<li>
<a href="https://openmendel.github.io/MendelBase.jl/#pedigree-file">Pedigree File</a>: Gives information about your individuals, such as name, parental information, and sex.</li>
<li>
<a href="https://openmendel.github.io/MendelBase.jl/#snp-definition-file">SNP Definition File</a>: Defines your SNPs with information such as SNP name, chromosome, position, and allele names.</li>
<li>
<a href="https://openmendel.github.io/MendelBase.jl/#snp-data-file">SNP Data File</a>: Holds the genotypes for your data set and must be a standard binary PLINK BED file in SNP major format. If you have a SNP data file, you must also have a SNP definition file.</li>
</ul>

<h3>
<a id="control-file" class="anchor" href="#control-file" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Control file<a id="control-file"></a>
</h3>

<p>The Control file is a text file consisting of keywords and their assigned values. The format of the Control file is:</p>

<pre><code>Keyword = Keyword_Value(s)
</code></pre>

<p>Below is an example of a simple Control file to run GWAS:</p>

<pre><code>#
# Input and Output files.
#
plink_input_basename = gwas 1 data
output_file = gwas 1 Output.txt
manhattan_plot_file = gwas 1 Manhattan Plot Output.png
#
# Analysis parameters for GWAS option.
#
regression = linear
regression_formula = Trait ~ Sex
</code></pre>

<p>In the example above, there are five keywords. The keyword <em>plink_input_basename</em> tells MendelGWAS that the input data files will comprise three <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/data.shtml">PLINK format</a> data files: <em>gwas 1 data.fam</em>, <em>gwas 1 data.bim</em>, and <em>gwas 1 data.bed</em>. The next two keywords specify the output files: <em>gwas 1 Output.txt</em> the results file - and <em>gwas 1 Manhattan Plot Output.png</em> - a plot of the results. The last two keywords specify analysis parameters. The text after the " = " are the keyword values.</p>

<h3>
<a id="keywords" class="anchor" href="#keywords" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Keywords<a id="keywords-table"></a>
</h3>

<p>This is a list of OpenMendel keywords relevant to GWAS. A list of OpenMendel keywords common to most analysis package can be found <a href="https://openmendel.github.io/MendelBase.jl/#keywords-table">here</a>. The names of keywords are <em>not</em> case sensitive. (The keyword <em>values</em> may be case sensitive.)</p>

<table>
<thead>
<tr>
<th>Keyword</th>
<th>Default Value</th>
<th>Allowed Values</th>
<th>Short Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>affected_designator</td>
<td>affected</td>
<td></td>
<td>Under logistic regression, the trait label assigned to cases</td>
</tr>
<tr>
<td>distribution</td>
<td></td>
<td>Binomial(), Gamma(), Normal(), Poisson(), etc.</td>
<td>Name of distribution function</td>
</tr>
<tr>
<td>link</td>
<td></td>
<td>LogitLink(), IdentityLink(), LogLink(), etc.</td>
<td>Name of link function</td>
</tr>
<tr>
<td>lrt_threshold</td>
<td>5e-8</td>
<td></td>
<td>Threshold for the score test p-value below which a likelihood ratio test is performed</td>
</tr>
<tr>
<td>maf_threshold</td>
<td>0.01</td>
<td></td>
<td>Threshold for the minor allele frequency below which SNPs are not analyzed</td>
</tr>
<tr>
<td>manhattan_plot_file</td>
<td></td>
<td></td>
<td>Name of file to hold Manhattan plot</td>
</tr>
<tr>
<td>regression</td>
<td></td>
<td>Linear, Logistic, or Poisson</td>
<td>Type of standard regression to perform; if left blank, the keywords distribution and link must be assigned values</td>
</tr>
<tr>
<td>regression_formula</td>
<td></td>
<td></td>
<td>Defines the regression model to analyze under the null hypothesis. See below for a description of the syntax to use</td>
</tr>
</tbody>
</table>

<p>The value for the keyword regression_formula takes the following form: the trait variable is separated from the predictors by " ~ ". For example, <code>regression_formula = Case_Control ~ Sex + BMI</code> means the trait labeled Case_Control will be modeled as a grand mean (intercept term) along with the covariates of Sex and BMI. Case_Control and BMI must be field names used in the pedigree file. (If you use a PLINK style FAM file, which does not use a header row, then the trait value should be referred to as "Trait".) Interactions between predictors are represented in the regression_formula using an "&amp;", for example, <code>Case_Control ~ Sex + BMI + Sex&amp;BMI</code> would add the interaction term to the previous main effects. (A product term, for example, Sex*BMI, is a short cut for main effects plus their interaction.) If the right hand side (after the " ~ ") is blank, then the model uses only the grand mean. Do not include a term for the SNPs as that is added automatically for each alternative model.</p>

<p>Under logistic regression, the cases are those individuals whose value at the trait field is the same as the label assigned to the keyword affected_designator. The trait field is the field listed on the left hand side (before the " ~ ") in the regression_formula. The controls are those individuals with non-missing values at the trait that are not cases. Of course individuals with missing values at the trait are neither cases nor controls. </p>

<h3>
<a id="data-files" class="anchor" href="#data-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Data Files</h3>

<p>GWAS requires a <a href="https://openmendel.github.io/MendelBase.jl/#control-file">Control file</a> and a <a href="https://openmendel.github.io/MendelBase.jl/#pedigree-file">Pedigree file</a>. Genotype data is provided in a <a href="https://openmendel.github.io/MendelBase.jl/#snp-data-file">SNP data file</a>, with a <a href="https://openmendel.github.io/MendelBase.jl/#snp-definition-file">SNP Definition File</a> describing the SNPs. OpenMendel will also accept <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/data.shtml">PLINK format</a> FAM and BIM files. Details on the format and contents of the Control and data files can be found on the <a href="https://openmendel.github.io/MendelBase.jl">MendelBase</a> documentation page. There are example data files in the GWAS <a href="https://github.com/OpenMendel/MendelGWAS.jl/tree/master/docs">docs</a> folder.</p>

<h3>
<a id="running-the-analysis" class="anchor" href="#running-the-analysis" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Running the Analysis</h3>

<p>To run this analysis package, first launch Julia. Then load the package with the command:</p>

<pre><code> julia&gt; using MendelGWAS
</code></pre>

<p>Next, if necessary, change to the directory containing your files, for example,</p>

<pre><code> julia&gt; cd("~/path/to/data/files/")
</code></pre>

<p>Finally, to run the analysis using the parameters in the control file Control_file.txt use the command:</p>

<pre><code> julia&gt; GWAS("Control_file.txt")
</code></pre>

<p><em>Note: The package is called</em> MendelGWAS <em>but the analysis function is called simply</em> GWAS.</p>

<h3>
<a id="interpreting-the-results" class="anchor" href="#interpreting-the-results" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Interpreting the results</h3>

<h3>
<a id="citation" class="anchor" href="#citation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Citation</h3>

<p>If you use this analysis package in your research, please cite the following reference in the resulting publications:</p>

<p><em>Lange K, Papp JC, Sinsheimer JS, Sripracha R, Zhou H, Sobel EM (2013) Mendel: The Swiss army knife of genetic analysis programs. Bioinformatics 29:1568-1570.</em></p>



<h3>
<a id="acknowledgments" class="anchor" href="#acknowledgments" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Acknowledgments</h3>

<p>This project is supported by the National Institutes of Health under NIGMS awards R01GM053275 and R25GM103774 and NHGRI award R01HG006139.</p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">MendelGWAS.jl maintained by <a href="https://github.com/OpenMendel">OpenMendel</a></p>
        <p>Published with <a href="https://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
